src/t8_schemes/t8_new_feature/t8_subelements/t8_subelements_quad_cxx.cxx

```diff
@@ -0,0 +1,1325 @@
+  binary_array[3] = binary_array_temp[2]; /* f0 <- f1 */
+
+  /* 3) use the rearranged binary representation, and the sub_id to determine the location of the subelement and store these information in an array */
+  /* 3.1) location[0] -> the face_number, the subelement is adjacent to */
+  /* 3.2) location[1] -> dummy, if the face is split or not */
+  /* 3.3) location[2] -> is the subelement the first or second subelement of the face (always the first, if the face is not split) */
+  T8_ASSERT (pquad_w_sub->subelement_id < pquad_w_sub->num_subelement_ids);
+
+  int sub_id = pquad_w_sub->subelement_id;
+  int sub_face_id = 1; /* initialize the sub_face_id as the second subelement of a face and change that if it is not true */
+  int face_number = 0;
+  int split;
+
+  int k = 0;
+
+  if (sub_id == 0) {
+    sub_face_id = 0; /* in this case, the subelement is the first element of the face */
+  }
+
+  while (sub_id >= 0 && k < P4EST_FACES) {
+    if (binary_array[k] == 1) {
+      sub_id -= 2; /* two subelements adjacent to face k */
+    }
+    else {
+      sub_id -= 1; /* one subelement adjacent to face k */
+    }
+    if (sub_id == 0) {
+      sub_face_id = 0; /* in this case, the subelement is the first element of the face */
+    }
+    k++;
+  }
+  face_number = k - 1;
+
+  if (binary_array[face_number] == 0) {
+    split = 0; /* the face is not split */ 
+  }
+  else {
+    split = 1; /* the face is split */
+  }
+
+  location[0] = face_number;
+  location[1] = split;
+  location[2] = sub_face_id;
+}
+
+t8_element_shape_t
+t8_default_scheme_sub_c::t8_element_get_shape (const t8_element_t * elem)
+{
+  const t8_quad_with_subelements *pquad_w_sub = (const t8_quad_with_subelements *) elem;
+
+  T8_ASSERT (t8_element_is_valid (elem));
+
+  if (pquad_w_sub->dummy_is_subelement != 1) { 
+    return T8_ECLASS_QUAD;
+  }
+  else { /* for quads, all subelements are triangles */
+    return T8_ECLASS_TRIANGLE;
+  }
+} 
+
+void
+t8_default_scheme_sub_c::t8_element_new (int length, t8_element_t ** elem)
+{
+  /* allocate memory for a quad with subelements */
+  t8_default_scheme_common_c::t8_element_new (length, elem);
+
+  int                 i;
+  for (i = 0; i < length; i++) {
+    t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem[i];
+    t8_element_init (1, elem[i], 0);
+    /* set dimension of quad to 2 */
+    T8_QUAD_SET_TDIM ((p4est_quadrant_t *) & pquad_w_sub->p4q, 2);
+  }
+}
+
+void
+t8_default_scheme_sub_c::t8_element_init (int length, t8_element_t * elem,
+                                          int new_called)
+{
+  t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem;
+
+  int                 i;
+
+  for (i = 0; i < length; i++) {
+    /* initalize subelement parameters */
+    pquad_w_sub[i].dummy_is_subelement = 0;
+    pquad_w_sub[i].subelement_type = -1;
+    pquad_w_sub[i].subelement_id = -1;
+    pquad_w_sub[i].num_subelement_ids = -1;
+
```

Comment:
I would like to use the new reset_subelement_values function to initialize the subelement parameters here. Then I realized that something is weird. We iterate over i from 0 to length but length from element_new is equal to 1. So why do we iterate here. Even if we call this function somewhere else with length > 1 this should not work since we are only using * elem instead of ** elem, right? In this case, casting etc. should look like in element_new I guess.